File: src/main/java/com/portofino/realtrainmodunofficial/network/SaveFormationPayload.java
package com.portofino.realtrainmodunofficial.network;

import com.portofino.realtrainmodunofficial.formation.TrainFormation;
import com.portofino.realtrainmodunofficial.formation.TrainFormationData;
import com.portofino.realtrainmodunofficial.item.TrainVehicleItem;
import com.portofino.realtrainmodunofficial.item.VehicleFormationItem;
import io.netty.buffer.ByteBuf;
import java.util.ArrayList;
import java.util.List;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.ItemStack;
import net.neoforged.neoforge.network.handling.IPayloadContext;

public record SaveFormationPayload(String name, List<String> vehicleIds, List<Boolean> reversedFlags) implements CustomPacketPayload {
   public static final CustomPacketPayload.Type<SaveFormationPayload> TYPE = new CustomPacketPayload.Type(ResourceLocation.fromNamespaceAndPath("realtrainmodunofficial", "save_formation"));
   public static final StreamCodec<ByteBuf, SaveFormationPayload> STREAM_CODEC;

   public CustomPacketPayload.Type<? extends CustomPacketPayload> type() {
      return TYPE;
   }

   public static void handleOnServer(SaveFormationPayload payload, IPayloadContext context) {
      context.enqueueWork(() -> {
         Player player = context.player();
         TrainFormation formation = new TrainFormation();
         formation.setName(payload.name());

         for(int i = 0; i < payload.vehicleIds().size(); ++i) {
            String vehicleId = (String)payload.vehicleIds().get(i);
            if (vehicleId != null && !vehicleId.isBlank()) {
               formation.addVehicle(vehicleId);
            }
         }

         for(InteractionHand hand : InteractionHand.values()) {
            ItemStack stack = player.getItemInHand(hand);
            if (stack.getItem() instanceof VehicleFormationItem || stack.getItem() instanceof TrainVehicleItem) {
               TrainFormationData.setFormation(stack, formation);
               break;
            }
         }

      });
   }

   static {
      STREAM_CODEC = StreamCodec.composite(ByteBufCodecs.STRING_UTF8, SaveFormationPayload::name, ByteBufCodecs.collection(ArrayList::new, ByteBufCodecs.STRING_UTF8), SaveFormationPayload::vehicleIds, ByteBufCodecs.collection(ArrayList::new, ByteBufCodecs.BOOL), SaveFormationPayload::reversedFlags, SaveFormationPayload::new);
   }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/network/SaveFormationPayload.kt
package cc.mirukuneko.realtrainmodrenewed.network

import cc.mirukuneko.realtrainmodrenewed.RealTrainModRenewed
import cc.mirukuneko.realtrainmodrenewed.formation.TrainFormation
import cc.mirukuneko.realtrainmodrenewed.formation.TrainFormationData
import cc.mirukuneko.realtrainmodrenewed.item.TrainVehicleItem
import cc.mirukuneko.realtrainmodrenewed.item.VehicleFormationItem
import io.netty.buffer.ByteBuf
import net.minecraft.network.codec.ByteBufCodecs
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.resources.Identifier
import net.minecraft.world.InteractionHand
import net.neoforged.neoforge.network.handling.IPayloadContext
import java.util.ArrayList

@JvmRecord
data class SaveFormationPayload(
    val name: String,
    val vehicleIds: List<String>,
    val reversedFlags: List<Boolean>,
) : CustomPacketPayload {
    override fun type(): CustomPacketPayload.Type<out CustomPacketPayload> = TYPE

    companion object {
        @JvmField
        val TYPE: CustomPacketPayload.Type<SaveFormationPayload> = CustomPacketPayload.Type(
            Identifier.fromNamespaceAndPath(RealTrainModRenewed.MODID, "save_formation")
        )

        @JvmField
        val STREAM_CODEC: StreamCodec<ByteBuf, SaveFormationPayload> = StreamCodec.composite(
            ByteBufCodecs.STRING_UTF8,
            { payload -> payload.name },
            ByteBufCodecs.collection({ ArrayList<String>() }, ByteBufCodecs.STRING_UTF8, 256),
            { payload -> payload.vehicleIds },
            ByteBufCodecs.collection({ ArrayList<Boolean>() }, ByteBufCodecs.BOOL, 256),
            { payload -> payload.reversedFlags },
            ::SaveFormationPayload,
        )

        @JvmStatic
        fun handleOnServer(payload: SaveFormationPayload, context: IPayloadContext) {
            context.enqueueWork {
                val player = context.player()
                val formation = TrainFormation()
                formation.name = payload.name

                for (vehicleId in payload.vehicleIds) {
                    if (vehicleId.isNotBlank()) {
                        formation.addVehicle(vehicleId)
                    }
                }

                for (hand in InteractionHand.entries) {
                    val stack = player.getItemInHand(hand)
                    if (stack.item is VehicleFormationItem || stack.item is TrainVehicleItem) {
                        TrainFormationData.setFormation(stack, formation)
                        break
                    }
                }
            }
        }
    }
}
